libs/langgraph/langgraph/pregel/__init__.py

```diff
@@ -1288,228 +1061,70 @@ async def astream(
-                    checkpoint = copy_checkpoint(checkpoint)
-                    for k in channels:
-                        if k in checkpoint["channel_versions"]:
-                            version = checkpoint["channel_versions"][k]
-                            checkpoint["versions_seen"][INTERRUPT][k] = version
-
+            async with AsyncPregelLoop(
+                input, config=config, checkpointer=self.checkpointer, graph=self
+            ) as loop:
+                aioloop = asyncio.get_event_loop()
                 # Similarly to Bulk Synchronous Parallel / Pregel model
                 # computation proceeds in steps, while there are channel updates
-                # channel updates from step N are only visible in step N+1,
+                # channel updates from step N are only visible in step N+1
                 # channels are guaranteed to be immutable for the duration of the step,
-                # channel updates being applied only at the transition between steps
-                stop = start + config["recursion_limit"] + 1
-                for step in range(start, stop):
-                    next_checkpoint, next_tasks = _prepare_next_tasks(
-                        checkpoint,
-                        processes,
-                        channels,
-                        managed,
-                        config,
-                        step,
-                        for_execution=True,
-                        manager=run_manager,
-                        get_next_version=(
-                            self.checkpointer.get_next_version
-                            if self.checkpointer
-                            else _increment
-                        ),
-                    )
-
-                    # assign pending writes to tasks
-                    if saved and saved.pending_writes:
-                        for task in next_tasks:
-                            task.writes.extend(
-                                (c, v)
-                                for tid, c, v in saved.pending_writes
-                                if tid == task.id
-                            )
-
-                    # if no more tasks, we're done
-                    if not next_tasks:
-                        if step == start:
-                            raise ValueError("No tasks to run in graph.")
-                        else:
-                            break
-
-                    # before execution, check if we should interrupt
-                    if _should_interrupt(
-                        checkpoint,
-                        interrupt_before,
-                        self.stream_channels_list,
-                        next_tasks,
-                    ):
-                        break
-                    else:
-                        checkpoint = next_checkpoint
-
+                # with channel updates applied only at the transition between steps
+                while loop.tick(
```

Comment:
And I guess loop.tick can be sync because we aren't currently doing io-bound work in it?

Reply:
yep, just python logic. all io things (like writing a checkpoint) get sent to background thread/task in both sync/async